README.md
raytracing
================

[![Codecov test
coverage](https://codecov.io/gh/salvatirehbein/raytracing/branch/master/graph/badge.svg)](https://codecov.io/gh/salvatirehbein/raytracing?branch=master)
[![CRAN
Downloads](http://cranlogs.r-pkg.org/badges/grand-total/raytracing?color=orange)](http://cran.r-project.org/package=raytracing)
[![DOI](https://zenodo.org/badge/295589682.svg)](https://zenodo.org/badge/latestdoi/295589682)
[![R build
status](https://github.com/salvatirehbein/raytracing/workflows/R-CMD-check/badge.svg)](https://github.com/salvatirehbein/raytracing/actions)

## Raytracing Documentation

The identification of the atmospheric Rossby wave ray paths is of
paramount importance for atmospheric scientists, climatologists,
meteorologists, and students in order to research, assign, monitor and
forecast weather and climate patterns. `raytracing` is designed to
detect Rossby waves trigger regions, characteristics, and paths, using a
zonally symmetric basic state.

It relies on the theory described in:

  - [Hoskins and Karoly
    (1981)](https://journals.ametsoc.org/jas/article/38/6/1179/20507/The-Steady-Linear-Response-of-a-Spherical)
  - [Karoly
    (1983)](https://www.sciencedirect.com/science/article/abs/pii/0377026583900131)
  - [Hoskins and Ambrizzi
    (1993)](https://journals.ametsoc.org/jas/article/50/12/1661/23207/Rossby-Wave-Propagation-on-a-Realistic)
  - [Yang and Hoskins
    (1996)](https://journals.ametsoc.org/jas/article/53/16/2365/24038/Propagation-of-Rossby-Waves-of-Nonzero-Frequency)

For a brief review, see Rehbein et al. (2020).

## Including Code

The `raytracing` include the below functions:

| Functions   | Arguments                                                                            | Description                                               |
| :---------- | :----------------------------------------------------------------------------------- | :-------------------------------------------------------- |
| betaks      | u, lat, lon, uname, ofile, a, plots, show.warnings                                   | Calculates Beta and Ks                                    |
| ray         | betam, u, lat, x0, y0, K, dt, itime, direction, interpolation, tl, a, verbose, ofile | Calculates the Rossby waves ray paths                     |
| ray\_path   | x, y                                                                                 | Calculate the ray paths / great circles                   |
| ray\_source | betam, u, lat, x0, y0, K, dt, itime, direction, interpolation, tl, a, verbose, ofile | Calculate the Rossby waves ray paths over a source region |
| trin        | y, yk, mercator                                                                      | Performs trigonometric interpolation                      |
| ypos        | y, lat, yk, mercator                                                                 | Interpolation selecting the nearest neighbor              |

**Table** **1** **-** `raytracing` functions, arguments, and its
description.

## Instalation

To install the CRAN version:

``` r
install.packages("raytracing")
```

To install the developing version:

``` r
remotes::install_github("salvatirehbein/raytracing")
```

## Example

Simple usage example, reproduced from [Coelho et
al. (2015)](https://link.springer.com/article/10.1007/s00382-015-2800-1).

``` r
library(raytracing)
input <- system.file("extdata",
                     "uwnd.mon.mean_200hPa_2014JFM.nc",
                     package = "raytracing")
b <- betaks(u = input)
rt <- ray_source(betam = b$betam,
                 u = b$u,
                 lat = b$lat,
                 K = 3,
                 itime = 10*4,
                 x0 = -c(130, 135),
                 y0 = -30,
                 dt = 6,
                 direction = -1,
                 interpolation = "trin")
```

    ## 
    ## K = 3  lat = -30   lon = -130 
    ## 
    ## K = 3  lat = -30   lon = -135

The `ray` or the `ray_source` functions return a `sf` `data.table` as
below:

``` r
head(rt)
```

    ## Simple feature collection with 6 features and 10 fields
    ## Geometry type: POINT
    ## Dimension:     XY
    ## Bounding box:  xmin: -130 ymin: -49.98866 xmax: -106.6441 ymax: -30
    ## Geodetic CRS:  WGS 84
    ##   K lat_ini lon_ini time iday       lat       lon lon_shift                  id
    ## 1 3     -30    -130    0    0 -30.00000 -130.0000  230.0000 K3_lati-30_loni-130
    ## 2 3     -30    -130    6    6 -32.47862 -128.5935  231.4065 K3_lati-30_loni-130
    ## 3 3     -30    -130   12   12 -35.64802 -126.2817  233.7183 K3_lati-30_loni-130
    ## 4 3     -30    -130   18   18 -39.65338 -122.2070  237.7930 K3_lati-30_loni-130
    ## 5 3     -30    -130   24    0 -44.48906 -115.7751  244.2249 K3_lati-30_loni-130
    ## 6 3     -30    -130   30    6 -49.98866 -106.6441  253.3559 K3_lati-30_loni-130
    ##   direction                    geometry
    ## 1        -1            POINT (-130 -30)
    ## 2        -1 POINT (-128.5935 -32.47862)
    ## 3        -1 POINT (-126.2817 -35.64802)
    ## 4        -1  POINT (-122.207 -39.65338)
    ## 5        -1 POINT (-115.7751 -44.48906)
    ## 6        -1 POINT (-106.6441 -49.98866)

``` r
tail(rt)
```

    ## Simple feature collection with 6 features and 10 fields (with 1 geometry empty)
    ## Geometry type: LINESTRING
    ## Dimension:     XY
    ## Bounding box:  xmin: -43.27348 ymin: -17.45744 xmax: -43.17391 ymax: -16.92884
    ## Geodetic CRS:  WGS 84
    ##     K lat_ini lon_ini time iday       lat       lon lon_shift
    ## 159 3     -30    -135  210   18 -17.45744 -43.27348  316.7265
    ## 160 3     -30    -135  216    0 -17.33805 -43.25002  316.7500
    ## 161 3     -30    -135  222    6 -17.22619 -43.22855  316.7714
    ## 162 3     -30    -135  228   12 -17.12113 -43.20884  316.7912
    ## 163 3     -30    -135  234   18 -17.02220 -43.19069  316.8093
    ## 164 3     -30    -135  240    0 -16.92884 -43.17391  316.8261
    ##                      id direction                       geometry
    ## 159 K3_lati-30_loni-135        -1 LINESTRING (-43.27348 -17.4...
    ## 160 K3_lati-30_loni-135        -1 LINESTRING (-43.25002 -17.3...
    ## 161 K3_lati-30_loni-135        -1 LINESTRING (-43.22855 -17.2...
    ## 162 K3_lati-30_loni-135        -1 LINESTRING (-43.20884 -17.1...
    ## 163 K3_lati-30_loni-135        -1 LINESTRING (-43.19069 -17.0...
    ## 164 K3_lati-30_loni-135        -1               LINESTRING EMPTY

``` r
# Large stationary wave numbers (Ks) or short wavelength are set to be Ks = 10
b$sfpoly$ksm <- ifelse(b$sfpoly$ksm > 10 & b$sfpoly$ksm <= 20, 10, 
                ifelse(b$sfpoly$ksm >=30 | b$sfpoly$ksm < 0, NA, 
                b$sfpoly$ksm))
# Plot Ks
plot(b$sfpoly["ksm"],
     lty = 0,
     axes = TRUE, 
     reset = FALSE,
     breaks = seq(0, 10, 2),
     pal= c("#f0ff00", "#ffce00", "#ff9a00", "#ff5a00", "#ff0000"),
     main = "Stationary Wave Number and Ray Tracings: JFM/2014")

# Plot contour maps
data(coastlines)
plot(coastlines, add = TRUE)
# Select linestrings for plotting only the lines
li <- rt[st_is(rt$geometry, "LINESTRING"), ]

# Plot ray traces
plot(li["lon_ini"],
    add = TRUE,
    lwd = 2,
    pal = colorRampPalette(c("black", "blue")))
```

![](README_files/figure-gfm/plot-1.png)<!-- -->

This figure is a reproduction of the Figure 9 from [Coelho et
al. (2015)](https://link.springer.com/article/10.1007/s00382-015-2800-1).

## citation

To cite raytracing in publications use this:

Rehbein, A., Ambrizzi, T., Ibarra-Espinosa, S., Dutra, L. M. M.: Rossby
Wave Ray Tracing v0.1.0. <https://github.com/salvatirehbein/raytracing>,
2020.

A BibTeX entry for LaTeX users is

    @Manual{ray,
        title = {raytracing: An R package for identification and tracking the atmospheric Rossby waves},
        author = {Amanda Rehbein and Tercio Ambrizzi and Sergio Ibarra-Espinosa and Livia M. M. Dutra},
        year = {2020},
        url = {https://github.com/salvatirehbein/raytracing},
      }

Thanks


NEWS.md
NEWS
===========
# TODO

- Review descriptions in order to substitute "source" by "departure point" or
  other
- Implement RWS function
- Implement WFA function
- Depreciate function ray
- Include function for plotting
- Export uwnd in beta and betaks with lat/lon in ascending order

### raytracing 0.5.0 (Date: 08 feb 2022)
- Implement function for filter by area of interest (wave_arrival)
- ray and ray_source functions now exports in csv instead of RDS.

### raytracing 0.4.0 (Date: 06 feb 2022)
- Substitute "expand.grid" by "data.frame" in ray_source function
- fix bugs in documentation

### raytracing 0.3.0 (Date: 22 aug 2021)
- Create betam, Ktotal, and Ks function from betaks
- Export netcdf files to be read by GrADS 

### raytracing 0.2.0 (Date: 23 dez 2020)
- Add correction for longitudes bigger than 180
- Add DOI with zenodo in README

### raytracing 0.1.1 (Date: 17 dez 2020)
- Add tryCatch to the ray_path function in order to avoid stopping
  when the wave does not propagate: [geo(i + 1)] problem
- Add sep = ";" in write.csv function at ray and ray_source functions
- Fix typo error in documentation of betaks function, replacing lat by lon

### raytracing 0.1.0 (Date: 22 out 2020)
- ray_source returns great circles
- betaks now returns spatial polygons
- trigonometric interpolation was added
- Preparing submission to CRAN

### raytracing 0.0.0.9000 (Date: 16 sep 2020)
- First news
- Preparing submission to CRAN and JOSS


R/beta_ks.R
#' Calculates Beta and Ks
#'
#' \code{betaks} ingests the time-mean zonal wind (u), transform it in
#' mercator coordinates (um); calculates the meridional gradient of
#' the absolute vorticity (beta) in mercator coordinates (betam);
#' and, finally, calculates stationary wavenumber (Ks) in mercator coordinates
#' (ksm) (see: Hoskins and Ambrizzi, 1993). \code{betaks} returns the um, betam,
#' and lat, for being ingested in \code{\link{ray}} or
#' \code{\link{ray_source}}.
#'
#' @param u String indicating the input data filename. The file to be
#' passed consists in a netCDF file with only time-mean zonal wind at one
#' pressure level, latitude in ascending order (not a requisite), and longitude
#' from 0 to 360.
#' It is required that the read dimensions express
#' longitude (in rows) x latitude (in columns).
#' \strong{u} also can be a numerical matrix with time-mean zonal wind at one
#' pressure level, latitude in ascending order (not a requisite), and longitude
#' from 0 to 360. It is required that the read dimensions express longitude
#' (in rows) x latitude (in columns).
#' @param ofile String indicating the file name for store output data.
#' If missing, will not return a netCDF file
#' @param uname String indicating the variable name field
#' @param lat String indicating the name of the latitude field. If
#' \strong{u} is a matrix, \strong{lat} must be numeric.
#' @param lon String indicating the name of the longitude field.If
#' \strong{u} is a matrix, \strong{lon} must be numeric from 0 to 360.
#' @param a Numeric indicating the Earth's radio (m)
#' @param plots Logical, if TRUE will produce filled.countour plots
#' @param show.warnings Logical, if TRUE will warns about NaNs in sqrt(<0)
#' @return list with one vector (lat) and 3 matrices (um, betam, and ksm)
#' @importFrom ncdf4 nc_open ncvar_get nc_close ncdim_def ncvar_def
#'  nc_create ncatt_put
#' @importFrom graphics filled.contour
#' @export
#' @examples {
#' # u is NetCDF and lat and lon characters
#' input <- system.file("extdata",
#'                      "uwnd.mon.mean_200hPa_2014JFM.nc",
#'                       package = "raytracing")
#' b <- betaks(u = input, plots = TRUE)
#' b$ksm[] <- ifelse(b$ksm[] >= 16 |
#'                   b$ksm[] <= 0, NA, b$ksm[])
#' cores <- c("#ff0000","#ff5a00","#ff9a00","#ffce00","#f0ff00")
#' graphics::filled.contour(b$ksm[, -c(1:5, 69:73)] ,
#'                          col = rev(colorRampPalette(cores, bias = 0.5)(20)),
#'                          main = "Ks")
#'
#' # u, lat and lon as numeric
#' input <- system.file("extdata",
#'                      "uwnd.mon.mean_200hPa_2014JFM.bin",
#'                       package = "raytracing")
#' u <- readBin(input,
#'              what = numeric(),
#'              size = 4,
#'              n = 144*73*4)
#' lat <- seq(-90, 90, 2.5)
#' lon <- seq(-180, 180 - 1, 2.5)
#' u <- matrix(u,
#'             nrow = length(lon),
#'             ncol = length(lat))
#' graphics::filled.contour(u, main = "Zonal Wind Speed [m/s]")
#' b <- betaks(u, lat, lon)
#' b$ksm[] <- ifelse(b$ksm[] >= 16 |
#'                   b$ksm[] <= 0, NA, b$ksm[])
#' cores <- c("#ff0000","#ff5a00","#ff9a00","#ffce00","#f0ff00")
#' graphics::filled.contour(b$ksm[, -c(1:5, 69:73)] ,
#'                          col = rev(colorRampPalette(cores, bias = 0.5)(20)),
#'                          main = "Ks")
#' }
betaks <- function(u,
                   lat = "lat",    # lon,
                   lon = "lon",
                   uname = "uwnd",   # lat
                   ofile,
                   a = 6371000,
                   plots = FALSE,
                   show.warnings = FALSE
) {
  if(inherits(u, "character")){
    message("Detecting NetCDF")
    ncu      <- ncdf4::nc_open(filename = u)
    u        <- ncdf4::ncvar_get(nc = ncu, varid = uname)
    lat      <- ncdf4::ncvar_get(nc = ncu, varid = lat)
    lon      <- ncdf4::ncvar_get(nc = ncu, varid = lon)
    ncdf4::nc_close(ncu)

  }


  dx       <- abs(lon[2] - lon[1])
  dy       <- abs(lat[2] - lat[1])
  nlat     <- length(lat)
  nlon     <- length(lon)

  # Define parameters and constants
  if(is.unsorted(lat)) {
    message("Sorting latitude from south to north")
    phi <- lat[order(lat, decreasing = FALSE)]
    u <- u[, nlat:1]
  } else {
    phi <- lat # nocov
  }

  if(plots) graphics::filled.contour(u, main = "u")

  omega <- 2*pi/(24*60*60)
  dphi <- dy*pi/180
  dphia <- dphi*a
  phirad <- matrix(phi*(pi/180),
                   nrow = 1,
                   ncol = nlat)

  # Calculate beta terms
  # beta (mercator) using eq 2.2 from Hoskins & Ambrizzi (1993)
  # using spherical coord. x cos phi
  # 1st term: df/dy ####
  dfdy <- matrix(NA,
                 nrow = 1,
                 ncol = nlat - 2)
  for(j in 2:(nlat - 1)) {
    dfdy[1, j-1] <- (
      2*omega*sin(phirad[1, j + 1]) - 2*omega*sin(phirad[1, j - 1]))/(2*dphia)
  }

  #  message("df/dy must be close to 0 at the poles and > 0 at the Equator\n")
  if(plots) graphics::plot(as.vector(dfdy),
                           main= "df/dy",
                           pch = 16)

  # 2nd term: d2u/dy2 ####
  m_phirad <- matrix(as.vector(phirad),
                     nrow = nlon,
                     ncol = nlat,
                     byrow = TRUE)
  uc <- u*cos(m_phirad)

  d2udy2 <- matrix(NA,
                   nrow = nlon,
                   ncol = nlat - 2)
  for(j in 2:(nlat - 1)) {
    for(i in 1:nlon) {
      d2udy2[i, j-1] <- (uc[i, j + 1] - 2*uc[i, j] + uc[i, j-1]) / (dphia^2)

    }
  }
  if(plots) graphics::filled.contour(d2udy2, main = expression("d2u/dy2"))

  # Calculate Beta ####
  m_dfdy <- matrix(as.vector(dfdy),
                   nrow = nlon,
                   ncol = nlat - 2,
                   byrow = TRUE)
  beta <- m_dfdy - d2udy2

  beta_f <- cbind(beta[, 1],
                  beta,
                  beta[, ncol(beta)])

  if(plots) graphics::filled.contour(beta_f[,], main = "Beta")

  # Calculate Beta Mercator --> beta * cos(phi) #####
  beta_mercator <- beta_f*cos(m_phirad)

  if(plots) graphics::filled.contour(beta_mercator,
                                     main = "Beta Mercator")

  # Ks Mercator #######
  # Ks (Mercator) using eq 2.13 from Hoskins & Ambrizzi (1993)
  ks_mercator <- matrix(NA, nrow = nlon, ncol = nlat)

  ks_mercator[] <- ifelse(
    beta_mercator[] < 0 & u[] != 0, -1,
    ifelse(
      beta_mercator[] > 0 & u[] < 0, 30,
      ifelse(
        beta_mercator[] == 0 | u[] == 0, 0,
        if(show.warnings) {
          a * sqrt( (beta_mercator * cos(m_phirad)) / u)
        } else {
          suppressWarnings(a * sqrt( (beta_mercator*cos(m_phirad)) / u))
        }
      )))

  ks_mercator[] <- ifelse(ks_mercator[] >= 16 &
                            ks_mercator[] != 30, 20, ks_mercator[])

  if(plots) graphics::hist(ks_mercator, main = "Ks")
  if(plots) graphics::filled.contour(ks_mercator, main = "Ks")

  # Adding sf objects ####
  # grid
  bb <- c(xmin = min(lon - 180),
          ymin = min(lat),
          xmax = max(lon - 180),
          ymax = max(lat))

  bb <- sf::st_bbox(bb)

  g <- sf::st_make_grid(sf::st_as_sfc(bb), n = c(nlon, nlat))

  # Coordinates in x (longitude) are organized from 0 to 360
  # To transform to a spatial object with lat/long coordinates
  # we need to shift the x coordinates to -180 to 180.
  # The matrix for `u`, `beta_mercator`, and `ks_mercator` are R objects with more
  # points on y (rows), i.e., those matrix have more rows than columns, or
  # longitudes are in y and latitudes on x.
  # The functions `filled.contour` and `image` read those transposed matrix,
  # adjusting them automatically on the plot, so we see longitude on x and
  # latitude on y.
  # To shift the spatial coordinates of the matrix we need to deal with the
  # rows of our matrix in R in the following way.

  u_sf <- rbind(u[(nrow(u)/2 + 1):nrow(u), ],
                u[1:(nrow(u)/2), ])

  betam_sf <- rbind(beta_mercator[(nrow(beta_mercator)/2 + 1):nrow(beta_mercator), ],
                    beta_mercator[1:(nrow(beta_mercator)/2), ])

  ksm_sf <- rbind(ks_mercator[(nrow(ks_mercator)/2 + 1):nrow(ks_mercator), ],
                  ks_mercator[1:(nrow(ks_mercator)/2), ])

  sfpoly <- sf::st_sf(data.frame(u = as.vector(u_sf)[1:length(g)],
                                 betam = as.vector(betam_sf)[1:length(g)],
                                 ksm = as.vector(ksm_sf)[1:length(g)]),
              geometry = g,
              crs = 4326)

  if(missing(ofile)){
    return(list(lat = phi,
                u = u,
                betam = beta_mercator,
                ksm = ks_mercator,
                sfpoly = sfpoly))
  } else {
    cat("Writting the netcdf in ",ofile,"...\n")

    # Write the netcdf ####
    # definition of dimensions
    XLONG <- ncdf4::ncdim_def(name = "longitude",
                              units = "degrees_east",
                              vals = lon)
    XLAT <- ncdf4::ncdim_def(name = "latitude",
                             units = "degrees_north",
                             vals = -lat)

    # definition of variables
    UWND <- ncdf4::ncvar_def(name = "uwnd",
                             units = "m s^-1",
                             dim = list(XLONG, XLAT),
                             longname="Zonal wind")
    BETAM <- ncdf4::ncvar_def(name = "betam",
                              units = "s^-1 m^-1",
                              dim = list(XLONG, XLAT),
                              longname="Meridional Gradient of the Absolute Vorticity in mercator coordinates")
    KSM <- ncdf4::ncvar_def(name = "Ks",
                            units = "",
                            dim = list(XLONG, XLAT),
                            longname = "Stationary Wavenumber (Ks)")
    vars_file <- ncdf4::nc_create(filename = ofile,
                                  vars = list(UWND, BETAM, KSM))

    cat(paste("The file has", vars_file$nvars, "variables\n"))
    cat(paste("The file has", vars_file$ndim, "dimensions\n"))

    # Global attribute to the file when varid = 0
    # otherwise write the variable name
    ncdf4::ncatt_put(nc = vars_file,
                     varid = 0, # 0 para o arquivo
                     attname = "title",
                     attval = "Basic state for calculate ray tracing")
    ncdf4::ncatt_put(nc = vars_file,
                     varid = 0, # 0 para o arquivo
                     attname = "Author",
                     attval = "")
    ncdf4::ncatt_put(nc = vars_file,
                     varid = 0, # 0 para o arquivo
                     attname = "institution",
                     attval = "Climate Group of Studies (GrEC)/University of Sao Paulo (USP)")
    ncdf4::ncatt_put(nc = vars_file,
                     varid = 0, # 0 para o arquivo
                     attname = "title",
                     attval = "Basic state for calculate ray tracing")
    ncdf4::ncatt_put(nc = vars_file,
                     varid = 0, # 0 para o arquivo
                     attname = "history",
                     attval = paste0("Created on ", Sys.time()))
    ncdf4::ncatt_put(nc = vars_file,
                     varid = 0, # 0 para o arquivo
                     attname = "references",
                     attval = "See: Hoskins and Ambrizzi (1993), Yang and Hoskins (1996), and Rehbein et al. (2020)")

    # Add variables to the ofile
    ncdf4::ncvar_put(nc = vars_file,
                     varid = UWND,
                     vals = u,
                     start = c(1,1),
                     count = c(nlon,nlat))
    ncdf4::ncvar_put(nc = vars_file,
                     varid = BETAM,
                     vals = beta_mercator,
                     start = c(1,1),
                     count = c(nlon,nlat))
    ncdf4::ncvar_put(nc = vars_file,
                     varid = KSM,
                     vals = ks_mercator,
                     start = c(1,1),
                     count = c(nlon,nlat))

    ncdf4::nc_close(nc = vars_file)

    # Returning a list with the calculated variables
    return(list(lat = phi,
                u = u,
                betam = beta_mercator,
                ksm = ks_mercator,
                sfpoly = sfpoly))

  }
}
